### PR TITLE
external_storage: move cloud code to separate cloud crate

### PR DESCRIPTION
**Pull request by gregwebs**
TODO:
- [X] merge [CloudDynamic kvproto upstream](https://github.com/pingcap/kvproto/pull/737)
- [x] merge [kvproto GRPC service definitions upstream](https://github.com/pingcap/kvproto/pull/748)

WIP:
- [X] test external_storage against S3
- [X] test external_storage against GCS
- [X] test kvproto change

### What problem does this PR solve?

We are organizing the cloud code in a separate crate. This will facilitate the cloud team being able to own the code and modifying builds to include or exclude cloud providers or in the future dynamically link them.

The encryption code reoganization is already merged here: https://github.com/tikv/tikv/pull/9647
This PR is for the external_storage code.

### What is changed and how it works?

* AWS and GCP code are moved from the external_storage crate into the cloud/aws and cloud/gcp crates
* components/external_storage/export is added to avoid cyclical dependencies. Cloud constructors of external_storage traits are imported here and then re-exported along with the external_storage crate for end users of external_storage
* rusoto_util code was already moved to aws/util in the encryption PR and in this PR we just delete the rusoto_util crate
* There is a flag `cloud-storage-grpc` for an experimental mode where external storage is served from a separate GRPC service over a unix domain socket.
* There is a flag `cloud-storage-dylib` for an experimental mode where external storage is called via C bindings. This code uses most of the same code from the GRPC flag. This code can be linked statically or loaded dynamically.

### Check List

Tests

- Unit test
- Manual test (WIP)

scli with localstack as s3. Tested both a text file (README) and a binary file (scli)

```
AWS_ACCESS_KEY_ID=test AWS_SECRET_ACCESS_KEY=test ./target/debug/examples/scli --storage S3 --endpoint http://localhost:4566 --region us-west-2 --bucket backup --name ext_storage --path . --file README.md save
aws s3 --endpoint http://localhost:4566 cp s3://backup/ext_storage .
diff ext_storage README.md
```

br with localstack as S3. Test dump a database, delete the database, then restore it. Restore uses same arguments as backup. Also test this after building with the flag `cloud-storage-dylib`.

```
AWS_ACCESS_KEY_ID=test AWS_SECRET_ACCESS_KEY=test ~/go/src/github.com/pingcap/br/bin/br --version backup full -s s3://backup/full2 --s3.endpoint http://localhost:4566 --s3.provider aws --s3.region us-west-2
```

br with fake-gcs server as gcs (uses some flags that I recently added on master).

```
br --gcs.endpoint http://storage.gcs.127.0.0.1.nip.io:4443/ --no-credentials --send-credentials-to-tikv=false --skip-check-path backup -s gcs://backup/alpha full
```

Delete the database from TiDB, after restore see that it exists again

```
br --gcs.endpoint http://storage.gcs.127.0.0.1.nip.io:4443/ --no-credentials --send-credentials-to-tikv=false restore -s gcs://backup/alpha full
```

Test cloud dynamic protobuf. Tested for AWS S3 as well. Tested setting both --host and --endpoint for GCS.
Below is the save command. Then test by doing the load command and performing a diff.

```
./target/debug/examples/scli --storage cloud --cloud-name gcp --host http://localhost:4443/ --bucket backup --name ext_storage9 --path . --file README.md save
```

Test experimental GRPC service. Run backup with binaries running that are built from these steps:

```
ENABLE_FEATURES=cloud-storage-grpc make build
./scripts/env cargo build -p external_storage_export --features=protobuf-codec,cloud-gcp,cloud-aws,cloud-storage-grpc && RUST_BACKTRACE=1 RUST_LIB_BACKTRACE=1 ./target/debug/tikv-cloud-storage
```

Dynamically loading the external storage code:

```
# manually edit the Cargo manifest to build a cdylib

# now build the lib
./scripts/env cargo build -p external_storage_export --features=cloud-storage-dylib,cloud-gcp,cloud-aws,protobuf-codec

# ensure the lib is int the working directory of tikv-server
cp target/debug/libexternal_storage_export.so .
```

If the lib cannot be dynamically loaded, you will see warnings during the backup in the logs.

Compatibilty
* Programs built from the TiKV lib (tikv-server, etc) now import the encryption_export crate instead of encryption directly.

### Release note

Move external_storage cloud provider code out of external_storage and into separate cloud provider crates.

### CODE REVIEW

File: components/external_storage/export/src/export.rs
Comment (lcwangchao) [inline components/external_storage/export/src/export.rs:38]: Any difference between BlobStorage and ExternalStorage?

File: components/external_storage/export/src/service.rs
Comment (lcwangchao) [inline components/external_storage/export/src/service.rs:24]: May be we can  handover the storage preparing works to grpc service itself. The grpc service can read configurations which contains a storage list  from a local file when it starts up.

File: components/external_storage/export/src/service.rs
Comment (lcwangchao) [inline components/external_storage/export/src/service.rs:56]: What other methods look such as writing data to a file? Maybe a multipart uploading interface should be exposed

File: components/external_storage/export/src/service.rs
Comment (gregwebs) [reply components/external_storage/export/src/service.rs:24]: This could work well for many use cases. And I think we can support this pre-loading. However, currently the br tool can specify an arbitrary backend configuration when it requests a backup.

File: components/external_storage/export/src/service.rs
Comment (gregwebs) [reply components/external_storage/export/src/service.rs:56]: yes, currently this is incomplete and I am looking into the upload/download part.

File: components/external_storage/export/src/export.rs
Comment (gregwebs) [reply components/external_storage/export/src/export.rs:38]: Currently there is no difference. This gives us some flexibility to change things in the future. But the real reason for duplicate traits is to avoid crates depending on each other. external_storage is not a dependency of cloud/* and cloud/* is not a dependency of external_storage. This crate, external_storage_export, binds them both together.

File: components/cloud/src/blob.rs
Comment (yiwu-arbug) [inline components/cloud/src/blob.rs:34]: put_object/get_object is more of "blob storage" than read/write?

File: components/cloud/src/blob.rs
Comment (yiwu-arbug) [inline components/cloud/src/blob.rs:114]: why not use empty string to represent none? That we don't need to introduce `StringNonEmpty`.

File: components/cloud/src/blob.rs
Comment (yiwu-arbug) [inline components/cloud/src/blob.rs:134]: Do not panic. return error instead.

File: components/cloud/src/blob.rs
Comment (yiwu-arbug) [inline components/cloud/src/blob.rs:109]: Can this be
```
pub struct BucketConf(backup::CloudDynamic);
```

File: components/cloud/src/blob.rs
Comment (gregwebs) [reply components/cloud/src/blob.rs:34]: Yes. Currently there is no difference with ExternalStorage so it made sense to keep the name the same until the API actually changes to better match blob storage.

File: components/cloud/src/blob.rs
Comment (gregwebs) [reply components/cloud/src/blob.rs:109]: I like having separate types for API and for domain modeling. protobuf types are very difficult to work with because we should be supporting both protobuf and prost and everything about their usage is different. There is an overhead to copying data around, but it doesn't matter for this metadata.

File: components/cloud/src/blob.rs
Comment (gregwebs) [reply components/cloud/src/blob.rs:114]: If an empty string represents none then every function must defensively check for an empty string. `StringNonEmpty` is a proof of an empty string: this check is done once and then non longer needs to be done anywhere else in the code. Making the caller prove data is already correct is a very powerful paradigm for reducing errors. Just the same as how Rust uses Option to prove that data is not null.

File: components/cloud/aws/src/kms.rs
Comment (gregwebs) [reply components/cloud/aws/src/kms.rs:17]: The encryption code already started using uppercase. I don't want to break backwards compatibility. The backup tools and code have tended towards lower case.

File: components/cloud/aws/src/s3.rs
Comment (gregwebs) [reply components/cloud/aws/src/s3.rs:1]: `git mv` doesn't matter. The issue is that in addition to the move some changes to the file have been made. In the past I used to try to break these things up into multiple commits but I don't think anyone ever looks at individual commits no matter how helpful it is.

Comment (gregwebs) [reply components/cloud/aws/src/s3.rs:1]: You can diff this with the git CLI rather than github. Then you have the ability to tweak the diff parameters. Note that you can already do `?w=` on github, but you can't adjust `-M --find-renames`

```
git diff -w -M45% origin/master gregwebs/cloud-storage
```

File: components/cloud/gcp/src/gcs.rs
Comment (gregwebs) [inline components/cloud/gcp/src/gcs.rs:1]: View this diff with

```
git diff -w -M40% origin/master gregwebs/cloud-storage
```

File: components/cloud/src/blob.rs
Comment (lcwangchao) [inline components/cloud/src/blob.rs:19]: Maybe we can make the io methods under BlobStorage async. It seems that async method are required for implementing the grpcio service and its better to usage **BlobStorage** trait directly. When tikv want use grcp to read/write file, it should use an object implemented the **ExternalStorage** by wrapping a grpcio client to make it.

Comment (gregwebs) [reply components/cloud/src/blob.rs:19]: This is implemented now in the latest commit. I did not convert the methods to be async because the compiler didn't like that idea, but I hope we will eventually.

File: components/external_storage/export/src/service.rs
Comment (gregwebs) [reply components/external_storage/export/src/service.rs:259]: The experimental GRPC service is completed now.

File: components/external_storage/export/src/service.rs
Comment (gregwebs) [reply components/external_storage/export/src/service.rs]: The GRPC API has changed a lot now.

File: components/cloud/aws/src/kms.rs
Comment (yiwu-arbug) [reply components/cloud/aws/src/kms.rs:17]: Can we have better naming to denote their usage?

File: components/cloud/src/blob.rs
Comment (yiwu-arbug) [reply components/cloud/src/blob.rs:34]: but "read/write" is confusing in the context of blob storage, if someday someone decide to use the blob storage API somewhere else.

File: components/cloud/aws/src/s3.rs
Comment (yiwu-arbug) [reply components/cloud/aws/src/s3.rs:1]: okay. cli is okay-ish but github has far better UI for side-by-side comparison and I can comment on it directly. 

Multiple commits would be especially helpful in this case though. They might not be useful other times.

Comment (yiwu-arbug) [reply components/cloud/aws/src/s3.rs:1]: I briefly check the diff of s3.rs and gcs.rs and don't think I have further comment.

**Review by yiwu-arbug — COMMENTED** [review]
Looking good to me. I'll approve after https://github.com/tikv/tikv/pull/9706#discussion_r607378285 is addressed.

File: components/cloud/aws/src/kms.rs
Comment (gregwebs) [reply components/cloud/aws/src/kms.rs:17]: One already has "KMS". Should the other have "backup"?

Comment (yiwu-arbug) [reply components/cloud/aws/src/kms.rs:17]: maybe `ENCRYPTION_AWS_VENDOR_NAME` and `BR_AWS_VENDOR_NAME`, and inline comment their usage?

File: components/cloud/src/blob.rs
Comment (gregwebs) [reply components/cloud/src/blob.rs:34]: Okay, I renamed to get/put

File: components/cloud/aws/src/kms.rs
Comment (gregwebs) [reply components/cloud/aws/src/kms.rs]: okay, done in https://github.com/tikv/tikv/pull/9706/commits/28cf7b7b9e813fc23c7a621ce2e7d6820cbb75c6

**Review by yiwu-arbug — APPROVED** [review]

File: components/external_storage/export/src/dylib.rs
Comment (lcwangchao) [inline components/external_storage/export/src/dylib.rs:28]: why should we keep a blob_storage in side this client? It seems the only use of the blob storage is to build the file path of shared data

File: components/external_storage/export/src/dylib.rs
Comment (gregwebs) [reply components/external_storage/export/src/dylib.rs:204]: I suspect the performance is not noticeable. Currently caching wouldn't do anything because a new ExternalStorageClient is created for each SST file so the functions are only called once.

File: components/external_storage/export/src/dylib.rs
Comment (gregwebs) [reply components/external_storage/export/src/dylib.rs:28]: yes, it is only used for metadata. I agree it should be removed. This code is tested to work but I still consider it experimental. We can spend more time cleaning things up if we commit to using this approach. Properly organizing the build for the cdylib is the first things I would look at cleaning up.

Comment (gregwebs) [reply components/external_storage/export/src/dylib.rs]: I went ahead and removed the blob_storage reference since it was an easy change to make.

File: components/external_storage/export/src/export.rs
Comment (kennytm) [inline components/external_storage/export/src/export.rs:77]: i'd suggest turning `BlobStore` to a generic type to avoid the doubling boxing.

```rust
struct BlobStore<B>(B);
impl<B: BlobStorage> ExternalStorage for BlobStore<B> { ... }
```

the `create_backend_inner` can use

```rust
Box::new(BlobStore(S3Storage::from_input(config.clone())?)))
```

File: components/cloud/src/blob.rs
Comment (kennytm) [inline components/cloud/src/blob.rs:174]:
```suggestion
        s.0
```

File: components/cloud/src/blob.rs
Comment (kennytm) [inline components/cloud/src/blob.rs:89]:
```suggestion
    fn error_required(msg: &str) -> io::Error {
```

File: components/cloud/src/blob.rs
Comment (kennytm) [inline components/cloud/src/blob.rs:97]: i'd rather you call it `static_` or `static_str` or `r#static`.

this name sounds like `stat(2)`.

File: components/external_storage/export/src/export.rs
Comment (gregwebs) [reply components/external_storage/export/src/export.rs:77]: okay. I have found it really easy to use Box but I will try to start using generics where possible.

File: Cargo.toml
Comment (gregwebs) [reply Cargo.toml:233]: This is on the checklist at the top of the PR.

File: components/cloud/aws/src/s3.rs
Comment (kennytm) [inline components/cloud/aws/src/s3.rs:193]: i don't know why we need that TODO comment. static credentials provided by the client is needed for non-cloud clusters.

**Review by kennytm — COMMENTED** [review]
rest LGTM. but i can't approve before that kvproto PR is merged.

File: components/external_storage/src/grpc_client.rs
Comment (overvenus) [inline components/external_storage/src/grpc_client.rs:51]: What about putting it into tikv's `data-dir` to prevent being deleted unexpectedly? 

Have you considered security for the approach?  Does it require mTLS?

File: components/external_storage/export/src/grpc_service.rs
Comment (overvenus) [inline components/external_storage/export/src/grpc_service.rs:30]: Is it in-process RPC?

Comment (gregwebs) [reply components/external_storage/export/src/grpc_service.rs:30]: No. For in process one would use the C dynamic lib.

File: components/external_storage/src/grpc_client.rs
Comment (gregwebs) [reply components/external_storage/src/grpc_client.rs:51]: yes, data-dir might be a a better place. This would operate on the same node, so TLS is less of a concern. More important for security would be securing access to the socket.

The GRPC API is experimental and I would prefer to remove it in the future and make the case for the C dynamic lib instead. So I would prefer not to spend more time on it now.

File: components/external_storage/export/src/grpc_service.rs
Comment (gregwebs) [reply components/external_storage/export/src/grpc_service.rs:30]: The idea behind RPC is that the server is ran as a separate process: thus the need for the socket.

**Pull request merged**
